src/test/java/io/aiven/kafka/connect/transforms/ConcatFieldsTest.java

```diff
@@ -0,0 +1,271 @@
+        assertEquals("Invalid value: null used for required field: \"bar\", schema type: STRING", e.getMessage());
+    }
+
+    @Test
+    void mapWithMissingField() {
+        final SinkRecord originalRecord = record(null, new HashMap<>());
+        assertDoesNotThrow(() -> transformation().apply(originalRecord),
+            FIELD + " field must be present and its value can't be null: " + originalRecord);
+    }
+
+    @Test
+    void mapWithoutSchema() {
+        final HashMap<Object, Object> valueMap = new HashMap<>();
+        valueMap.put(TEST_FIELD, TEST_VALUE);
+        valueMap.put(FOO_FIELD, FOO_VALUE);
+        valueMap.put(BAR_FIELD, BAR_VALUE);
+        valueMap.put(AGE_FIELD, AGE_VALUE);
+        final HashMap<Object, Object> newValueMap = new HashMap<>();
+        newValueMap.put(TEST_FIELD, TEST_VALUE);
+        newValueMap.put(FOO_FIELD, FOO_VALUE);
+        newValueMap.put(BAR_FIELD, BAR_VALUE);
+        newValueMap.put(AGE_FIELD, AGE_VALUE);
+        newValueMap.put(FIELD, TEST_VALUE.toString() + "-" + FOO_VALUE + "-" + BAR_VALUE + "-" + AGE_VALUE);
+
+        final SinkRecord originalRecord = record(null, valueMap);
+        final SinkRecord transformedRecord = transformation().apply(originalRecord);
+        assertEquals(setNewValue(originalRecord, null, newValueMap), transformedRecord);
+    }
+
+    @Test
+    void mapWithoutSchemaMissingField() {
+        final HashMap<Object, Object> valueMap = new HashMap<>();
+        valueMap.put(TEST_FIELD, TEST_VALUE);
+        valueMap.put(BAR_FIELD, BAR_VALUE);
+        valueMap.put(AGE_FIELD, AGE_VALUE);
+        final HashMap<Object, Object> newValueMap = new HashMap<>();
+        newValueMap.put(TEST_FIELD, TEST_VALUE);
+        newValueMap.put(BAR_FIELD, BAR_VALUE);
+        newValueMap.put(AGE_FIELD, AGE_VALUE);
+        newValueMap.put(FIELD, TEST_VALUE.toString() + "-*-" + "Baz" + "-" + AGE_VALUE);
+
+        final SinkRecord originalRecord = record(null, valueMap);
+        final SinkRecord transformedRecord = transformation().apply(originalRecord);
+        assertEquals(setNewValue(originalRecord, null, newValueMap), transformedRecord);
+    }
+
+    @Test
+    void mapWithSchema() {
+        final HashMap<Object, Object> valueMap = new HashMap<>();
+        valueMap.put(TEST_FIELD, TEST_VALUE);
+        valueMap.put(FOO_FIELD, FOO_VALUE);
+        valueMap.put(BAR_FIELD, BAR_VALUE);
+        valueMap.put(AGE_FIELD, AGE_VALUE);
+        final HashMap<Object, Object> newValueMap = new HashMap<>();
+        newValueMap.put(TEST_FIELD, TEST_VALUE);
+        newValueMap.put(FOO_FIELD, FOO_VALUE);
+        newValueMap.put(BAR_FIELD, BAR_VALUE);
+        newValueMap.put(AGE_FIELD, AGE_VALUE);
+        newValueMap.put(FIELD, TEST_VALUE.toString() + "-" + FOO_VALUE + "-" + BAR_VALUE + "-" + AGE_VALUE);
+
+        final SinkRecord originalRecord = record(VALUE_SCHEMA, valueMap);
+        final SinkRecord transformedRecord = transformation().apply(originalRecord);
```

Comment:
I'm afraid the Key variant effectively doesn't work or works incorrectly. If you run the Key tests, `transformedRecord` will have the value overwritten with what was supposed to be the new key. Below go some pointers.